File: crates/agenthub-core/src/models/adapter.rs
//! Read-only adapter route and configuration-plan DTOs.

use serde::{Deserialize, Serialize};

use super::AgentId;

/// Persisted connection table selected by an adapter analysis request.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum AdapterSourceKind {
    Account,
    Provider,
}

impl AdapterSourceKind {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Account => "account",
            Self::Provider => "provider",
        }
    }

    pub fn parse(value: &str) -> Option<Self> {
        match value.trim().to_ascii_lowercase().as_str() {
            "account" => Some(Self::Account),
            "provider" => Some(Self::Provider),
            _ => None,
        }
    }
}

/// Product bucket for Adapter page tabs. Orthogonal to [`AdapterSourceKind`]
/// (table origin) and [`AdapterRoute`] (projection).
///
/// Derived from [`super::AdapterCredentialClass`] at apply time:
/// API Key → `api`, OAuth shapes → `oauth`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum AdapterProfileMode {
    Api,
    Oauth,
}

impl AdapterProfileMode {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Api => "api",
            Self::Oauth => "oauth",
        }
    }

    pub fn parse(value: &str) -> Option<Self> {
        match value.trim().to_ascii_lowercase().as_str() {
            "api" => Some(Self::Api),
            "oauth" => Some(Self::Oauth),
            _ => None,
        }
    }

    /// Map a classified credential family onto the persisted `mode` bucket.
    ///
    /// `Unknown` cannot become a profile: classify already fails closed.
    pub fn from_credential_class(class: super::AdapterCredentialClass) -> Option<Self> {
        match class {
            super::AdapterCredentialClass::ApiKey => Some(Self::Api),
            super::AdapterCredentialClass::OauthAuthJson
            | super::AdapterCredentialClass::OauthOther => Some(Self::Oauth),
            super::AdapterCredentialClass::Unknown => None,
        }
    }
}

/// Input to the read-only route analysis service. `source_id` is always a DB id.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AdapterRouteRequest {
    pub source_kind: AdapterSourceKind,
    pub source_id: String,
    pub target_agent_id: AgentId,
}

/// The only routes a compatibility rule can return.
///
/// Keeping this closed prevents UI callers from treating an unknown route as
/// safe to apply.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AdapterRoute {
    ConfigSync,
    NativeEndpoint,
    LocalBridge,
    Unsupported,
}

impl AdapterRoute {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::ConfigSync => "config_sync",
            Self::NativeEndpoint => "native_endpoint",
            Self::LocalBridge => "local_bridge",
            Self::Unsupported => "unsupported",
        }
    }

    pub fn parse(value: &str) -> Option<Self> {
        match value.trim().to_ascii_lowercase().as_str() {
            "config_sync" => Some(Self::ConfigSync),
            "native_endpoint" => Some(Self::NativeEndpoint),
            "local_bridge" => Some(Self::LocalBridge),
            "unsupported" => Some(Self::Unsupported),
            _ => None,
        }
    }

    pub fn is_profile_supported(self) -> bool {
        !matches!(self, Self::Unsupported)
    }
}

/// Lifecycle state for a persisted adapter profile.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AdapterProfileStatus {
    Applying,
    Active,
    NeedsAttention,
}

impl AdapterProfileStatus {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Applying => "applying",
            Self::Active => "active",
            Self::NeedsAttention => "needs_attention",
        }
    }

    pub fn parse(value: &str) -> Option<Self> {
        match value.trim().to_ascii_lowercase().as_str() {
            "applying" => Some(Self::Applying),
            "active" => Some(Self::Active),
            "needs_attention" => Some(Self::NeedsAttention),
            _ => None,
        }
    }
}

/// Persisted, credential-free record of an adapter configuration projection.
///
/// The profile only identifies its source and any generated provider; it never
/// stores source credentials or generated configuration content.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AdapterProfile {
    pub id: String,
    pub name: String,
    pub source_kind: AdapterSourceKind,
    pub source_id: String,
    pub target_agent_id: AgentId,
    pub route: AdapterRoute,
    /// Product tab: API conversion vs OAuth proxy. Independent of `route`.
    pub mode: AdapterProfileMode,
    pub status: AdapterProfileStatus,
    pub rule_id: String,
    pub rule_version: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub generated_provider_id: Option<String>,
    /// Actual loopback port for a local bridge. This is intentionally absent
    /// until the runtime has successfully bound its listener.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub local_port: Option<u16>,
    /// Whether a local bridge profile should be restored by its desktop host.
    /// The value is persisted even while a profile is waiting for recovery so
    /// the host can make a deterministic startup decision.
    pub auto_start: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub last_error_code: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

/// Optional, typed filters for persisted adapter profiles.
///
/// This is deliberately a pure DTO: repository implementations remain the
/// only place that turns it into SQL, keeping GUI/Tauri callers from building
/// their own filter semantics.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AdapterProfileFilter {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source_kind: Option<AdapterSourceKind>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub target_agent_id: Option<AgentId>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mode: Option<AdapterProfileMode>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub route: Option<AdapterRoute>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub status: Option<AdapterProfileStatus>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub auto_start: Option<bool>,
}

/// Confidence and availability of an adapter rule.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum AdapterSupport {
    Stable,
    Experimental,
    Unsupported,
}

/// Planner-facing implementation maturity of a graph edge.
///
/// Distinct from [`AdapterSupport`] (matrix cell confidence) and from
/// [`AdapterApplyPlan::can_apply`] (whether a write can happen *now*).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum AdapterMaturity {
    /// Matrix cell open + [`AdapterSupport::Stable`].
    Stable,
    /// Matrix cell open + [`AdapterSupport::Experimental`].
    Experimental,
    /// Recorded cell with gates closed, or explain-only (e.g. Codex → Claude).
    Preview,
    /// No edge / Other / unsupported fallback.
    #[default]
    None,
}

impl AdapterMaturity {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Stable => "stable",
            Self::Experimental => "experimental",
            Self::Preview => "preview",
            Self::None => "none",
        }
    }

    pub fn parse(value: &str) -> Option<Self> {
        match value.trim().to_ascii_lowercase().as_str() {
            "stable" => Some(Self::Stable),
            "experimental" => Some(Self::Experimental),
            "preview" => Some(Self::Preview),
            "none" => Some(Self::None),
            _ => None,
        }
    }
}

/// Structured presentation / gate class for analyze UI (not a write authorization).
///
/// UI must prefer this over parsing `reason` text. Write permission remains
/// `AdapterApplyPlan.can_apply` only.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AdapterGateKind {
    /// Applicable or ordinary preview with no special gate chrome.
    #[default]
    None,
    /// Stable/experimental rule that is intentionally preview-only.
    PreviewOnly,
    /// Closed experimental subscription-bridge candidate (e.g. Codex OAuth → Claude).
    SubscriptionCandidate,
    /// Generic missing / unsupported combination.
    Unsupported,
}

impl AdapterGateKind {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::None => "none",
            Self::PreviewOnly => "preview_only",
            Self::SubscriptionCandidate => "subscription_candidate",
            Self::Unsupported => "unsupported",
        }
    }

    pub fn parse(value: &str) -> Option<Self> {
        match value.trim().to_ascii_lowercase().as_str() {
            "none" => Some(Self::None),
            "preview_only" => Some(Self::PreviewOnly),
            "subscription_candidate" => Some(Self::SubscriptionCandidate),
            "unsupported" => Some(Self::Unsupported),
            _ => None,
        }
    }
}

/// Product-facing reuse path derived from the public route and credential.
///
/// This is presentation only; the domain route remains [`AdapterRoute`].
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AdapterReusePath {
    ApiEndpoint,
    NativeSubscription,
    LocalBridge,
    #[default]
    None,
}

/// A safe, structured description of one required future action.
///
/// `secret = true` is a reference to the selected Connection only. Such an
/// action must never carry `value`, so neither analysis nor plan can leak a
/// credential.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AdapterAction {
    pub kind: String,
    pub target: String,
    pub description: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub value: Option<String>,
    pub secret: bool,
}

/// Verifiable official source for a compatibility rule.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AdapterEvidence {
    pub label: String,
    pub url: String,
    pub verified_at: String,
}

/// Safe route preview. This intentionally contains no credentials or raw configs.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AdapterRouteAnalysis {
    pub route: AdapterRoute,
    pub support: AdapterSupport,
    pub reason: String,
    pub actions: Vec<AdapterAction>,
    pub limitations: Vec<String>,
    pub evidence: Vec<AdapterEvidence>,
    /// Capability-matrix rule id when a cell matched; never a secret.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub rule_id: Option<String>,
    /// Structured gate class for UI; prefer over parsing `reason`.
    #[serde(default)]
    pub gate_kind: AdapterGateKind,
}

/// Whether an eventual apply would need a local runtime service.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AdapterServiceImpact {
    None,
    RequiresLocalBridge,
}

/// A single configuration field that an eventual apply would write.
///
/// `secret = true` intentionally has no `value`: it says to reference the
/// selected Connection when apply is introduced, not to expose the secret.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AdapterPlanChange {
    pub target: String,
    pub field: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub value: Option<String>,
    pub secret: bool,
}

/// Safe apply preview. `plan()` is the only public planner exit.
///
/// `can_apply` is true only when **both** hold:
/// 1. the capability matrix cell is open (`can_apply` + all gates), and
/// 2. plan's private `write_gate` allows a write *now* (bind implementation
///    exists and the secret is resolvable for this ticket `source_kind`).
/// The matrix alone never authorizes writes. Maturity describes the edge;
/// `can_apply` describes today's write path.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AdapterApplyPlan {
    pub analysis: AdapterRouteAnalysis,
    pub target_agent_id: AgentId,
    pub can_apply: bool,
    /// Four-tier edge maturity. Independent of `can_apply`.
    #[serde(default)]
    pub maturity: AdapterMaturity,
    /// Product-facing reuse path derived from the public route.
    #[serde(default)]
    pub reuse_path: AdapterReusePath,
    /// Planner-facing reason. Same gist as `analysis.reason`.
    #[serde(default)]
    pub reason: String,
    pub service_impact: AdapterServiceImpact,
    pub changes: Vec<AdapterPlanChange>,
}

/// Write request for a supported adapter route.  Unlike [`AdapterApplyPlan`],
/// this is executed only by the narrow apply service.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AdapterApplyRequest {
    pub source_kind: AdapterSourceKind,
    pub source_id: String,
    pub target_agent_id: AgentId,
}

/// Persisted outcome of applying an adapter route.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AdapterApplyResult {
    pub profile: AdapterProfile,
    /// The generated provider is safe to return: it contains a reference
    /// marker, never the source secret.
    pub provider: super::Provider,
}

File: src/lib/backend/contracts/adapter.ts
import type { AgentKey } from '@/lib/types';
import retryableErrorContract from './retryable-error-contract.json';

/** Saved connection table selected for a read-only adapter route preview. */
export type AdapterSourceKind = 'account' | 'provider';

export interface AdapterRouteRequest {
  sourceKind: AdapterSourceKind;
  /** Database row id; never a label, credential, or config body. */
  sourceId: string;
  targetAgentId: AgentKey;
}

/** Closed compatibility routes; unknown wire values are rejected at the adapter boundary. */
export type AdapterRoute = 'native_endpoint' | 'local_bridge' | 'config_sync' | 'unsupported';

/** User-facing reuse family derived from the planner route. */
export type AdapterReusePath = 'api_endpoint' | 'native_subscription' | 'local_bridge' | 'none';

/**
 * Persisted credential family for a profile. Orthogonal to `sourceKind` (table) and `route`
 * (projection). API Key conversions stay `api` even when they use a local protocol bridge.
 */
export type AdapterProfileMode = 'api' | 'oauth';

/** A rule can be stable, experimental, or explicitly unsupported. */
export type AdapterSupport = 'stable' | 'experimental' | 'unsupported';

/**
 * Planner four-tier maturity of a graph edge.
 * Distinct from `support` (matrix confidence) and `canApply` (write now).
 */
export type AdapterMaturity = 'stable' | 'experimental' | 'preview' | 'none';

/**
 * Structured gate / presentation class from core analyze.
 * Prefer this over parsing `reason` text. Does not authorize writes.
 */
export type AdapterGateKind =
  | 'none'
  | 'preview_only'
  | 'subscription_candidate'
  | 'unsupported';

export type AdapterActionKind =
  | 'set_config'
  | 'set_env'
  | 'reference_connection_secret'
  | 'requires_local_bridge';

/** Safe action preview. A secret action only refers to the selected Connection. */
export type AdapterAction =
  | {
      kind: AdapterActionKind;
      target: string;
      description: string;
      value?: string;
      secret: false;
    }
  | {
      kind: AdapterActionKind;
      target: string;
      description: string;
      secret: true;
      value?: never;
    };

/** Official, date-stamped basis for the compatibility conclusion. */
export interface AdapterEvidence {
  label: string;
  url: string;
  verifiedAt: string;
}

/** Safe core response: no secret values or raw configuration are present. */
export interface AdapterRouteAnalysis {
  route: AdapterRoute;
  support: AdapterSupport;
  reason: string;
  actions: AdapterAction[];
  limitations: string[];
  evidence: AdapterEvidence[];
  /** Capability-matrix rule id when a cell matched. */
  ruleId?: string | null;
  /** Structured gate class for UI chrome; prefer over parsing reason. */
  gateKind?: AdapterGateKind;
}

export type AdapterServiceImpact = 'none' | 'requires_local_bridge';

/** One future config mutation. A secret change has no serializable value. */
export type AdapterPlanChange =
  | { target: string; field: string; value?: string; secret: false }
  | { target: string; field: string; secret: true; value?: never };

/** Safe preview of the eventual configuration mutation. `plan()` is the only planner exit. */
export interface AdapterApplyPlan {
  analysis: AdapterRouteAnalysis;
  targetAgentId: AgentKey;
  /**
   * True only when the capability matrix is open **and** plan's private
   * `write_gate` allows a write now (Account stays false). Matrix alone
   * never authorizes writes.
   */
  canApply: boolean;
  /** Four-tier edge maturity. Missing on older wires → treat as `none`. */
  maturity?: AdapterMaturity;
  /** Planner-derived presentation path. Missing on older wires → derive from route. */
  reusePath?: AdapterReusePath;
  /**
   * Planner-facing reason. Prefer this over `analysis.reason` when
   * `canApply` is false (same-edge Account adds an explicit write-gate note).
   */
  reason?: string;
  serviceImpact: AdapterServiceImpact;
  changes: AdapterPlanChange[];
}

/** Persisted lifecycle state for a credential-free adapter projection. */
export type AdapterProfileStatus = 'applying' | 'active' | 'needs_attention';

/** A saved adapter projection. It identifies connections but never serializes their secrets. */
export interface AdapterProfile {
  id: string;
  name: string;
  sourceKind: AdapterSourceKind;
  sourceId: string;
  targetAgentId: AgentKey;
  route: Exclude<AdapterRoute, 'unsupported'>;
  /** Credential family: API Key conversion vs official-login proxy. Independent of `route`. */
  mode: AdapterProfileMode;
  status: AdapterProfileStatus;
  ruleId: string;
  ruleVersion: string;
  generatedProviderId?: string | null;
  /** Bound loopback port for a local bridge; direct routes leave this empty. */
  localPort?: number | null;
  /** Restored by the desktop host after launch when this is a local bridge. */
  autoStart: boolean;
  lastErrorCode?: string | null;
  createdAt: string;
  updatedAt: string;
}

/** Credential-free, in-process state of one local bridge listener. */
export type AdapterBridgeRuntimeState =
  | 'starting'
  | 'running'
  | 'stopping'
  | 'stopped'
  | 'error'
  | 'degraded';

export type AdapterBridgeUpstreamStatus =
  | 'unknown'
  | 'connected'
  | 'stopped'
  | 'degraded'
  | 'unavailable';

/** One inbound local-route request. Never includes Authorization, bodies, or keys. */
export interface AdapterBridgeInboundRequest {
  at: string;
  method: string;
  path: string;
  status: number;
  ok: boolean;
}

export type RouteTraceStageStatus = 'pending' | 'ok' | 'failed' | 'skipped';

export interface RouteTraceMember {
  label: string;
  sourceKind: string;
  sourceId: string;
  ticketId?: string | null;
}

export interface RouteTracePoolAttempt {
  member: RouteTraceMember;
  status: RouteTraceStageStatus;
  code?: string | null;
  message?: string | null;
}

export interface RouteTraceLocalAuth {
  status: RouteTraceStageStatus;
  profileId?: string | null;
  port?: number | null;
  code?: string | null;
  message?: string | null;
}

export interface RouteTracePool {
  status: RouteTraceStageStatus;
  selectedMember?: RouteTraceMember | null;
  attempts?: RouteTracePoolAttempt[];
  code?: string | null;
  message?: string | null;
}

export interface RouteTraceConversion {
  status: RouteTraceStageStatus;
  path: string;
  result?: string | null;
  code?: string | null;
  message?: string | null;
}

export interface RouteTraceUpstreamAuth {
  status: RouteTraceStageStatus;
  httpStatus?: number | null;
  code?: string | null;
  message?: string | null;
}

export interface RouteTraceUpstream {
  status: RouteTraceStageStatus;
  url?: string | null;
  member?: RouteTraceMember | null;
  model?: string | null;
  upstreamModel?: string | null;
  httpStatus?: number | null;
  code?: string | null;
  message?: string | null;
}

/** One completed local-route request trace for monitoring. Credential-free. */
export interface AdapterBridgeRouteTrace {
  requestId: string;
  at: string;
  profileId?: string | null;
  method: string;
  path: string;
  httpStatus: number;
  ok: boolean;
  model?: string | null;
  latencyMs?: number | null;
  ttftMs?: number | null;
  inputTokens?: number | null;
  outputTokens?: number | null;
  localAuth: RouteTraceLocalAuth;
  pool: RouteTracePool;
  conversion: RouteTraceConversion;
  upstreamAuth: RouteTraceUpstreamAuth;
  upstream: RouteTraceUpstream;
  failureStage?: string | null;
}

/** Loopback listener status. `localToken` is the bearer that actually authenticates. */
export interface AdapterBridgeRuntimeStatus {
  profileId: string;
  state: AdapterBridgeRuntimeState;
  port?: number | null;
  endpoint?: string | null;
  startedAt?: string | null;
  upstreamStatus?: AdapterBridgeUpstreamStatus | string | null;
  /** Newest first. Empty when no tool has connected yet. */
  recentInbound?: AdapterBridgeInboundRequest[];
  /** Newest first. Per-request route traces for monitoring. */
  recentRouteTraces?: AdapterBridgeRouteTrace[];
  /** Authenticated inbound requests since this process started (not ring-capped). */
  totalRequestCount?: number;
  /** Failed authenticated inbound requests since this process started. */
  failedRequestCount?: number;
  /** ISO time of the latest authenticated inbound request this process has seen. */
  lastRequestAt?: string | null;
  /** Loopback bearer (`ahb_…`) accepted by this listener. */
  localToken?: string | null;
}

export type AdapterApplyRequest = AdapterRouteRequest;

/** Generated provider is safe: it contains a connection reference, never a credential value. */
export interface AdapterApplyResult {
  profile: AdapterProfile;
  provider: import('@/lib/types').Provider;
}

export interface AdapterProfileFilter {
  sourceKind?: AdapterSourceKind;
  sourceId?: string;
  targetAgentId?: AgentKey;
  mode?: AdapterProfileMode;
  route?: Exclude<AdapterRoute, 'unsupported'>;
  status?: AdapterProfileStatus;
  autoStart?: boolean;
}

/** Structured Adapter command error shared by Tauri and mock. */
export interface AdapterCommandErrorFields {
  code: string;
  message: string;
  details?: string | null;
  retryable: boolean;
}

export class AdapterCommandError extends Error implements AdapterCommandErrorFields {
  readonly code: string;
  readonly details?: string | null;
  readonly retryable: boolean;

  constructor(fields: AdapterCommandErrorFields) {
    super(fields.message);
    this.name = 'AdapterCommandError';
    this.code = fields.code;
    this.details = fields.details ?? null;
    this.retryable = fields.retryable;
  }
}

/** Shared with desktop `is_adapter_error_retryable` via retryable-error-contract.json. */
export function isAdapterErrorCodeRetryable(code: string): boolean {
  if (retryableErrorContract.retryableExact.some((item) => item === code)) return true;
  return retryableErrorContract.retryablePrefixes.some((prefix) => code.startsWith(prefix));
}

export function adapterCommandError(fields: {
  code: string;
  message: string;
  details?: string | null;
  retryable?: boolean;
}): AdapterCommandError {
  return new AdapterCommandError({
    code: fields.code,
    message: fields.message,
    details: fields.details ?? null,
    retryable: fields.retryable ?? isAdapterErrorCodeRetryable(fields.code),
  });
}

export type RoutePoolSurface = 'messages' | 'responses' | 'chat_completions';
export type RoutePoolDialect = 'claude' | 'codex' | 'grok' | 'kimi' | 'dsh' | 'generic';

export type MemberAvailability = 'ready' | 'cooling' | 'isolated' | 'disabled';

/** Credential-free member row. Never includes login secrets or Hub tokens. */
export interface RouteMemberOverview {
  id?: string;
  sourceKind: AdapterSourceKind;
  sourceId: string;
  /** Safe account/provider label; never a credential or source-id fallback. */
  displayLabel?: string;
  /** Masked OAuth refresh-token tail, e.g. `**1234`; never the raw token. */
  refreshTokenTail?: string;
  enabled: boolean;
  priority?: number;
  availability?: MemberAvailability;
}

/** Credential-free default-pool overview. Never includes `hubToken`. */
export interface DefaultRoutePoolOverview {
  id: string;
  targetAgentId: AgentKey;
  surface: RoutePoolSurface;
  dialect: RoutePoolDialect;
  unifiedGatewayEnrolled: boolean;
  gatewayPort?: number | null;
  members: RouteMemberOverview[];
  listedModels?: string[];
}

export interface DefaultRoutePoolList {
  enabled: boolean;
  pools: DefaultRoutePoolOverview[];
  /** Kimi and DSH share one chat-completions local token when true. */
  chatCompletionsShared?: boolean;
}

export interface AttachPoolOwnedAuthorizationRequest {
  sourceKind: AdapterSourceKind;
  sourceId: string;
  targetAgentId: AgentKey;
  surface: RoutePoolSurface;
}

/** One credential-free Connections row selected for route-pool enrollment. */
export interface SyncConnectionSource {
  sourceKind: AdapterSourceKind;
  sourceId: string;
}

/** Optional selection for route-pool enrollment; omitted means all eligible rows. */
export interface SyncConnectionAuthorizationsRequest {
  sources: SyncConnectionSource[];
}

export interface SyncConnectionAuthorizationsResult {
  added: number;
  skipped: number;
}

/** Shared local-gateway status for the board switch. */
export type LocalGatewayStatus = {
  running: boolean;
  port: number | null;
  statuses: AdapterBridgeRuntimeStatus[];
  /** Local-auth failures without a bound route (newest first). */
  unauthenticatedTraces?: AdapterBridgeRouteTrace[];
  /** True while restore or start is bringing local forwarding back. */
  restarting: boolean;
};

/** Loopback bearer for the tokens page. */
export type LocalTokenRecord = {
  id: string;
  poolId: string;
  token: string;
  name: string;
  primary: boolean;
};

/** Result of a tokens-page model-path probe with an entry key. */
export type LocalTokenProbeOutcome =
  | 'ok'
  | 'unauthorized'
  | 'unreachable'
  | 'rejected'
  | 'invalid';

export type SourceModelCatalogSource = 'live' | 'custom' | 'empty';

export type SourceModelCatalog = {
  models: string[];
  source: SourceModelCatalogSource;
  canCustomize: boolean;
};

export type LocalTokenProbeResult = {
  outcome: LocalTokenProbeOutcome;
  httpStatus: number | null;
  latencyMs: number;
  upstreamStatus: string | null;
  requestUrl: string | null;
  requestMethod: string | null;
  requestBody: string | null;
  responseBody: string | null;
  errorMessage: string | null;
};

export interface AdapterPort {
  analyze(request: AdapterRouteRequest): Promise<AdapterRouteAnalysis>;
  plan(request: AdapterRouteRequest): Promise<AdapterApplyPlan>;
  listProfiles(filter?: AdapterProfileFilter): Promise<AdapterProfile[]>;
  listDefaultRoutePools(): Promise<DefaultRoutePoolList>;
  listLocalTokens(): Promise<LocalTokenRecord[]>;
  listLocalTokenModels(token: string): Promise<string[]>;
  refreshLocalTokenModels(token: string): Promise<string[]>;
  setLocalTokenCustomModels(token: string, models: string[]): Promise<string[]>;
  ensureSourceModelCatalog(
    sourceKind: AdapterSourceKind,
    sourceId: string,
  ): Promise<SourceModelCatalog>;
  setSourceCustomModels(
    sourceKind: AdapterSourceKind,
    sourceId: string,
    models: string[],
  ): Promise<SourceModelCatalog>;
  testLocalToken(
    endpoint: string,
    token: string,
    path: string,
    model?: string | null,
  ): Promise<LocalTokenProbeResult>;
  setLocalToken(poolId: string, token: string): Promise<LocalTokenRecord>;
  createLocalToken(poolId: string, name: string): Promise<LocalTokenRecord>;
  setLocalTokenName(id: string, name: string): Promise<LocalTokenRecord>;
  deleteLocalToken(id: string): Promise<void>;
  setChatCompletionsShared(shared: boolean): Promise<DefaultRoutePoolList>;
  attachPoolOwnedAuthorization(
    request: AttachPoolOwnedAuthorizationRequest,
  ): Promise<DefaultRoutePoolOverview>;
  setRouteAuthorizationEnabled(
    sourceKind: AdapterSourceKind,
    sourceId: string,
    enabled: boolean,
  ): Promise<number>;
  setRouteAuthorizationPriority(
    sourceKind: AdapterSourceKind,
    sourceId: string,
    priority: number,
  ): Promise<number>;
  removeRouteAuthorization(
    sourceKind: AdapterSourceKind,
    sourceId: string,
  ): Promise<number>;
  recycleRouteMembership(
    sourceKind: AdapterSourceKind,
    sourceId: string,
  ): Promise<number>;
  syncConnectionAuthorizations(
    request?: SyncConnectionAuthorizationsRequest,
  ): Promise<SyncConnectionAuthorizationsResult>;
  enrollNativeToGateway(profileId: string): Promise<DefaultRoutePoolOverview>;
  apply(request: AdapterApplyRequest): Promise<AdapterApplyResult>;
  remove(profileId: string): Promise<void>;
  startBridge(profileId: string): Promise<AdapterBridgeRuntimeStatus>;
  stopBridge(profileId: string): Promise<AdapterBridgeRuntimeStatus>;
  getBridgeStatus(profileId: string): Promise<AdapterBridgeRuntimeStatus>;
  setBridgeAutoStart(profileId: string, autoStart: boolean): Promise<AdapterProfile>;
  startLocalGateway(): Promise<LocalGatewayStatus>;
  stopLocalGateway(): Promise<LocalGatewayStatus>;
  getLocalGatewayStatus(): Promise<LocalGatewayStatus>;
}
